R/tar_download.R

```diff
@@ -0,0 +1,104 @@
+#' @title Download a file from a remote source, checking for changes
+#'   first
+#' @description Create a target that downnloads a file if it has changed since
+#'   the last attempt, based responds to a change
+#'   in an arbitrary value. If the value changes, the target reruns.
+#' @details `tar_download()` creates a pair of targets, one upstream
+#'   and one downstream, usingn `tar_change()`. always runs and returns
+#'   header information (eTag and modified time, if available), to check
+#'   if the remote value has changed. This header gets referenced in the
+#'   downstream target, which causes the downstream target to download the
+#'   if the header information changes. The downstream target has "file" format
+#'   and stores the location of te downloaded file The behavior is cancelled if
+#'   `cue` is `tar_cue(depend = FALSE)` or `tar_cue(mode = "never")`.
+#' @export
+#' @return A list of two targets, one upstream and one downstream.
+#'   The upstream one checks url headers, and the downstream one
+#'   downloads the file to it. See the examples for details.
+#' @param url The URL to fetch the file from
+#' @param destfile Whether to invoke tidy evaluation
+#' @param destdir the directory to download the file to 
+#' @param stop_on_no_internet Whether to fail in the absence of an internet connection.
+#' Default is `FALSE` and only produces a warning
+#' @param handle a handle created by `curl::new_handle()` to use when fetching the file
+#'   or checking headers. Useful for authentication.
+#' @param ... other parameters passed to `tar_change()`
+#' @examples
+#' \dontrun{
+#' 
+# targets::tar_dir({
+# targets::tar_script({
+#   library(tarchetypes)
+#   tar_pipeline(
+#     tar_download(gitfile, url = "https://github.com/wlandau/tarchetypes/archive/master.zip")
+#   )
+# })
+# targets::tar_make(callr_function = NULL)
+# })
+#' }
+tar_download <- function(
+  name,
+  url,
+  destfile = basename(url),
+  destdir = ".",
+  handle = NULL,
+  stop_on_no_internet = FALSE,
+  ...
+) {
+  if(!requireNamespace("curl", quietly = TRUE)) {
+    stop("tar_download() requires the the package 'curl' to be installed")
+  }
+  handle <- handle %||% curl::new_handle()
+  tar_change(
+    name,
+    command = substitute({
+      tar_download_file(url, destfile, destdir, stop_on_no_internet, handle)
+      }, env = list(url = url, destfile = destfile, destdir = destdir, stop_on_no_internet = stop_on_no_internet, handle = handle)),
+    change = substitute({
+      tar_url_hash(url, stop_on_no_internet, handle)
+      },  env = list(url = url, stop_on_no_internet = stop_on_no_internet, handle = handle)),
+    format = "file"
+    ...
+  )
+}
+
+#' @rdname tar_download
+#' @export
+tar_download_file <- function(url, destfile, destdir, stop_on_no_internet = FALSE, handle = curl::new_handle())
+  {
+  if (!curl::has_internet()) {
```

Comment:
Other archetypes might require an internet connection, so it may be nice to put this in a new `assert_internet()` in `R/utils_assert.R`.

Reply:
Your assertion patterns don't currently allow for "warning only" options, which I want here. (The use-case being that in the absence of a connection the workflow should be able to continue with current versions of files.)  Do you want to incorporate that into your assertion setup or should I just do something conditional here?

Reply:
I see. How about a `check_internet()` utility?

```r
validate_internet <- function(assert_internet = FALSE) {
  trn(assert_internet, suggest_internet(), assert_internet())
}

suggest_internet <- function(msg = NULL) {
  if (!curl::has_internet()) {
    warn_validate("no internet")
  }
}

warn_validate <- function(...) {
  warning(warning_validate(...))
}

warning_validate <- function(...) {
  structure(
    list(message = paste0(..., collapse = ""), call = NULL),
    class = c(
      "condition_validate",
      "condition_tarchetypes",
      "warning",
      "condition"
    )
  )
}
```

Reply:
I still need to test for internet to determine whether to cancel the target.  Should this return a logical value?

Reply:
Honestly for me atm this seems like a way to turn 5 lines of readable code into 20 harder-to-understand lines.

Reply:
I just updated the code above to include a call to `tar_cancel()`. 

There are more functions, but they help us avoid nested if/else logic, and I think each function might be reusable for other archetypes. And `warn_validate()` and `warning_validate()` support custom conditions for warnings, which make exception handling and testing nicer.